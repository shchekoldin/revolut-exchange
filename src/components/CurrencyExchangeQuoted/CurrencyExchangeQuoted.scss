@import "~revolut/styles/_vars";
@import "~revolut/styles/_mixins";

.CurrencyExchangeQuoted {
    position: relative;
    flex: 1;
    padding: 2.5em 1.5em 0;
    box-sizing: border-box;
    min-width: 100%;

    .currencyAmount {
        overflow: hidden;

        .label {
            font-size: 40px;
            float: left;
            margin-top: 2px;
        }

        .amount {
            font-size: 40px;
            text-align: right;
            float: right;
            color: $font-color;
            width: 60%;
        }
    }

    .sub {
        margin-top: 0.5rem;
        font-size: 0.9em;
        color: lighten($body-color, 20%);
        clear: both;
        font-weight: bold;

        .currencyBalance {
            float: left;
        }

        .rate {
            float: right;
        }
    }
}
