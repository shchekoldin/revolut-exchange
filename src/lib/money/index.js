// @flow
function format(value: number, precision: number = 2) {
    return value.toFixed(precision);
}

function formatRate(value: number) {
    return format(value, 4);
}

export default {
    format,
    formatRate,
};
