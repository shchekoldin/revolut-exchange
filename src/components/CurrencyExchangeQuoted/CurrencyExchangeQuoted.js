// @flow
import React from 'react';
import i18n from 'revolut/i18n';
import money from 'revolut/lib/money';
import styles from './CurrencyExchangeQuoted.scss';

type Props = {
    baseCurrency: string,
    quoteCurrency: string,
    quoteRate: number,
    balance: number,
    baseAmount: number,
};

class CurrencyExchangeQuoted extends React.Component<Props> {
    render() {
        const {
            baseCurrency,
            quoteCurrency,
            quoteRate,
            balance,
            baseAmount,
        } = this.props;
        const baseRate = 1 / quoteRate;
        const quoteAmount = baseAmount * quoteRate;

        return (
            <div className={styles.CurrencyExchangeQuoted}>
                <div>
                    <div className={styles.currencyAmount}>
                        <div className={styles.label}>{quoteCurrency}</div>
                        {(quoteAmount > 0) ? (
                            <div className={styles.amount}>+{money.format(quoteAmount, 1)}</div>
                        ) : null}
                    </div>
                    <div className={styles.sub}>
                        <div className={styles.currencyBalance}>
                            <div>You have: {i18n.currency[quoteCurrency]}{money.format(balance)}</div>
                        </div>
                        <div className={styles.rate}>
                            <span>1{i18n.currency[quoteCurrency]}</span>
                            <span>=</span>
                            <span>{i18n.currency[baseCurrency]}{money.format(baseRate)}</span>
                        </div>
                    </div>
                </div>
            </div>
        );
    }
}

export default CurrencyExchangeQuoted;
