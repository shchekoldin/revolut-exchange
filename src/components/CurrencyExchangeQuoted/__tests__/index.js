import React from 'react';
import {shallow} from 'enzyme';
import chai from 'chai';
import money from 'revolut/lib/money';
import CurrencyExchangeQuoted from '../CurrencyExchangeQuoted';

describe('Component <CurrencyExchangeQuoted /> should render without errors', () => {
    it('render with static props', () => {
        const props = {
            baseCurrency: 'EUR',
            quoteCurrency: 'GBP',
            quoteRate: 0.5,
            balance: 100,
            baseAmount: 50,
        };

        const wrapper = shallow(<CurrencyExchangeQuoted {...props}/>);

        chai.expect(wrapper.instance()).to.be.not.undefined();
        chai.expect(wrapper.find('.currencyAmount').find('.label').text()).to.equal(props.quoteCurrency);
        chai.expect(wrapper.find('.currencyAmount').find('.amount').text()).to.equal(
            `+${money.format(props.baseAmount * props.quoteRate, 1)}`,
        );

        expect(wrapper).toMatchSnapshot();
    });
});
