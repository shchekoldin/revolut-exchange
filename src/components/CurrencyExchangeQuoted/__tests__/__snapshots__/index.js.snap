// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component <CurrencyExchangeQuoted /> should render without errors render with static props 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CurrencyExchangeQuoted
    balance={100}
    baseAmount={50}
    baseCurrency="EUR"
    quoteCurrency="GBP"
    quoteRate={0.5}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div>
        <div
          className="currencyAmount"
        >
          <div
            className="label"
          >
            GBP
          </div>
          <div
            className="amount"
          >
            +
            25.0
          </div>
        </div>
        <div
          className="sub"
        >
          <div
            className="currencyBalance"
          >
            <div>
              You have: 
              £
              100.00
            </div>
          </div>
          <div
            className="rate"
          >
            <span>
              1
              £
            </span>
            <span>
              =
            </span>
            <span>
              €
              2.00
            </span>
          </div>
        </div>
      </div>,
      "className": "CurrencyExchangeQuoted",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="currencyAmount"
          >
            <div
              className="label"
            >
              GBP
            </div>
            <div
              className="amount"
            >
              +
              25.0
            </div>
          </div>,
          <div
            className="sub"
          >
            <div
              className="currencyBalance"
            >
              <div>
                You have: 
                £
                100.00
              </div>
            </div>
            <div
              className="rate"
            >
              <span>
                1
                £
              </span>
              <span>
                =
              </span>
              <span>
                €
                2.00
              </span>
            </div>
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="label"
              >
                GBP
              </div>,
              <div
                className="amount"
              >
                +
                25.0
              </div>,
            ],
            "className": "currencyAmount",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "GBP",
                "className": "label",
              },
              "ref": null,
              "rendered": "GBP",
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "+",
                  "25.0",
                ],
                "className": "amount",
              },
              "ref": null,
              "rendered": Array [
                "+",
                "25.0",
              ],
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="currencyBalance"
              >
                <div>
                  You have: 
                  £
                  100.00
                </div>
              </div>,
              <div
                className="rate"
              >
                <span>
                  1
                  £
                </span>
                <span>
                  =
                </span>
                <span>
                  €
                  2.00
                </span>
              </div>,
            ],
            "className": "sub",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <div>
                  You have: 
                  £
                  100.00
                </div>,
                "className": "currencyBalance",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "You have: ",
                    "£",
                    "100.00",
                  ],
                },
                "ref": null,
                "rendered": Array [
                  "You have: ",
                  "£",
                  "100.00",
                ],
                "type": "div",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <span>
                    1
                    £
                  </span>,
                  <span>
                    =
                  </span>,
                  <span>
                    €
                    2.00
                  </span>,
                ],
                "className": "rate",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      "1",
                      "£",
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    "1",
                    "£",
                  ],
                  "type": "span",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "=",
                  },
                  "ref": null,
                  "rendered": "=",
                  "type": "span",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      "€",
                      "2.00",
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    "€",
                    "2.00",
                  ],
                  "type": "span",
                },
              ],
              "type": "div",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div>
          <div
            className="currencyAmount"
          >
            <div
              className="label"
            >
              GBP
            </div>
            <div
              className="amount"
            >
              +
              25.0
            </div>
          </div>
          <div
            className="sub"
          >
            <div
              className="currencyBalance"
            >
              <div>
                You have: 
                £
                100.00
              </div>
            </div>
            <div
              className="rate"
            >
              <span>
                1
                £
              </span>
              <span>
                =
              </span>
              <span>
                €
                2.00
              </span>
            </div>
          </div>
        </div>,
        "className": "CurrencyExchangeQuoted",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="currencyAmount"
            >
              <div
                className="label"
              >
                GBP
              </div>
              <div
                className="amount"
              >
                +
                25.0
              </div>
            </div>,
            <div
              className="sub"
            >
              <div
                className="currencyBalance"
              >
                <div>
                  You have: 
                  £
                  100.00
                </div>
              </div>
              <div
                className="rate"
              >
                <span>
                  1
                  £
                </span>
                <span>
                  =
                </span>
                <span>
                  €
                  2.00
                </span>
              </div>
            </div>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <div
                  className="label"
                >
                  GBP
                </div>,
                <div
                  className="amount"
                >
                  +
                  25.0
                </div>,
              ],
              "className": "currencyAmount",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "GBP",
                  "className": "label",
                },
                "ref": null,
                "rendered": "GBP",
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "+",
                    "25.0",
                  ],
                  "className": "amount",
                },
                "ref": null,
                "rendered": Array [
                  "+",
                  "25.0",
                ],
                "type": "div",
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <div
                  className="currencyBalance"
                >
                  <div>
                    You have: 
                    £
                    100.00
                  </div>
                </div>,
                <div
                  className="rate"
                >
                  <span>
                    1
                    £
                  </span>
                  <span>
                    =
                  </span>
                  <span>
                    €
                    2.00
                  </span>
                </div>,
              ],
              "className": "sub",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <div>
                    You have: 
                    £
                    100.00
                  </div>,
                  "className": "currencyBalance",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      "You have: ",
                      "£",
                      "100.00",
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    "You have: ",
                    "£",
                    "100.00",
                  ],
                  "type": "div",
                },
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <span>
                      1
                      £
                    </span>,
                    <span>
                      =
                    </span>,
                    <span>
                      €
                      2.00
                    </span>,
                  ],
                  "className": "rate",
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": Array [
                        "1",
                        "£",
                      ],
                    },
                    "ref": null,
                    "rendered": Array [
                      "1",
                      "£",
                    ],
                    "type": "span",
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "=",
                    },
                    "ref": null,
                    "rendered": "=",
                    "type": "span",
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": Array [
                        "€",
                        "2.00",
                      ],
                    },
                    "ref": null,
                    "rendered": Array [
                      "€",
                      "2.00",
                    ],
                    "type": "span",
                  },
                ],
                "type": "div",
              },
            ],
            "type": "div",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
